File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

  static char[][] board;
  static int[][] count;
  static int[] di = {0, 0, 1, -1};
  static int[] dj = {1, -1, 0, 0};
  static int n, m;

  public static void main(String[] args) throws IOException {

    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());

    board = new char[n][m];
    count = new int[n][m];

    for (int i = 0; i < n; i++) {
      String s = br.readLine();
      board[i] = s.toCharArray();
    }

    bfs();

  }

  private static void bfs() {

    Queue<Node> queue = new LinkedList<>();
    queue.offer(new Node(0, 0));
    count[0][0] = 1;

    while (!queue.isEmpty()) {
      Node cur = queue.poll();

      for (int d = 0; d < 4; d++) {
        int nx = cur.i + di[d];
        int ny = cur.j + dj[d];

        if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
          if (board[nx][ny] == '1' && count[nx][ny] == 0) {
            queue.offer(new Node(nx, ny));
            count[nx][ny] = count[cur.i][cur.j] + 1;
          }
        }
      }
    }

    System.out.println(count[n - 1][m - 1]);
  }

  static class Node {
    int i;
    int j;

    Node(int i, int j) {
      this.i = i;
      this.j = j;
    }
  }

}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.kt
fun main() {
    val (n, m) = readln().split(" ").map(String::toInt)
    val board = Array(n) { readln().toCharArray() }
    val count = Array(n) { IntArray(m) }
    val queue = ArrayDeque<IntArray>()
    queue.addLast(intArrayOf(0, 0))

    val dx = arrayOf(1, -1, 0, 0)
    val dy = arrayOf(0, 0, 1, -1)

    count[0][0] = 1
    while (queue.isNotEmpty()) {
        val cur = queue.removeFirst()

        for (i in 0..3) {
            val nx = cur[0] + dx[i]
            val ny = cur[1] + dy[i]

            if (nx in 0 until n && ny in 0 until m) {
                if (board[nx][ny] == '1' && count[nx][ny] == 0) {
                    queue.add(intArrayOf(nx, ny))
                    count[nx][ny] = count[cur[0]][cur[1]] + 1
                }
            }
        }
    }

    print(count[n - 1][m - 1])

}
